Source/API/Extras/CBLIncrementalStore.m

```diff
@@ -2144,102 +2150,41 @@ - (void) mergeContextsWithUpdatedIDs: (NSArray*)updatedIDs deletedIDs: (NSArray*
-                NSManagedObject *destObj = [mObj valueForKey:relName];
-                if (!destObj || destObj.isFault || [destObj hasFaultForRelationshipNamed:invRel.name])
-                    continue;
-                
-                [needRefreshObjects addObject: destObj];
-            }    
-        }
 
-        if (updated.count > 0)
-            [userInfo setObject: updated forKey: NSUpdatedObjectsKey];
-        if (inserted.count > 0)
-            [userInfo setObject: inserted forKey: NSInsertedObjectsKey];
-    }
+            NSMutableSet *objectsToRefresh = [NSMutableSet set];
+            [objectsToRefresh addObjectsFromArray:insertedIDs.allObjects];
+            [objectsToRefresh addObjectsFromArray:updatedIDs.allObjects];
+            [objectsToRefresh addObjectsFromArray:deletedIDs.allObjects];
+            [objectsToRefresh addObjectsFromArray:refreshedIDs.allObjects];
 
-    if (deletedIDs.count > 0) {
-        NSMutableArray* deleted = [NSMutableArray arrayWithCapacity: deletedIDs.count];
-        for (NSManagedObjectID* moID in deletedIDs) {
-            NSManagedObject* mObj = [strongRootContext objectWithID: moID];
-            [deleted addObject: mObj];
-            [updatedEntities addObject: mObj.entity.name];
-        }
-        [userInfo setObject: deleted forKey: NSDeletedObjectsKey];
-    }
-    
-    // Clear cache:
-    for (NSString* entity in updatedEntities) {
-        [self purgeCachedObjectsForEntityName: entity];
-    }
-    
-    // Firing fault:
-    if (self.observingManagedObjectContexts.count > 1) {
-        for (NSManagedObjectID *moID in [needRefreshObjects allObjects]) {
-            [self fireFaultOnManagedObjectID: moID];
-        }
-    }
-    
-    NSNotification* notification =
-        [NSNotification notificationWithName: NSManagedObjectContextObjectsDidChangeNotification
-                                      object: strongRootContext
-                                    userInfo: userInfo];
-    [strongRootContext mergeChangesFromContextDidSaveNotification: notification];
-
-    for (NSManagedObjectContext *context in self.observingManagedObjectContexts) {
-        if (context != strongRootContext) {
-            [context performBlock:^{
-                [context mergeChangesFromContextDidSaveNotification: notification];
-            }];
-        }
-    }
-}
+            INFO(@"Will refresh %@ / %@ objects", @(objectsToRefresh.count), @(context.registeredObjects.count));
 
-- (void) fireFaultOnManagedObjectID: (NSManagedObjectID *)moID {
-    for (NSManagedObjectContext *context in self.observingManagedObjectContexts) {
-        if (context == self.rootContext)
-            continue;
-        
-        [context performBlock:^{
-            NSManagedObject *mObj = [context objectRegisteredForID: moID];
-            if (!mObj) return;
-            
-            INFO(@"firing fault : %@ in %@ on %@", moID, context, [NSThread currentThread]);
-            [mObj willAccessValueForKey: nil];
```

Comment:
I understand that calling willAccessValueForKey: with nil will make sure that the object will be firing fault.

From Apple Doc of the willAccessValueForKey: method :
> You can invoke this method with the key value of nil to ensure that a fault has been fired, as illustrated by the following example.
>
> [aManagedObject willAccessValueForKey:nil];